src/MSALWrapper/AuthFlow/IntegratedWindowsAuthentication.cs

```diff
@@ -0,0 +1,170 @@
+                            this.errors)
+                            .ConfigureAwait(false);
+                        tokenResult.SetAuthenticationType(AuthType.Silent);
+
+                        return new AuthFlowResult(tokenResult, this.errors);
+                    }
+                    catch (MsalUiRequiredException ex)
+                    {
+                        this.errors.Add(ex);
+                        this.logger.LogDebug($"Silent auth failed, re-auth is required.\n{ex.Message}");
+                    }
+                }
+                catch (MsalServiceException ex)
+                {
+                    this.logger.LogWarning($"MSAL Service Exception! (Not expected)\n{ex.Message}");
+                    this.errors.Add(ex);
+                }
+                catch (MsalClientException ex)
+                {
+                    this.logger.LogWarning($"Msal Client Exception! (Not expected)\n{ex.Message}");
+                    this.errors.Add(ex);
+                }
+                catch (NullReferenceException ex)
+                {
+                    this.logger.LogWarning($"Msal unexpected null reference! (Not Expected)\n{ex.Message}");
+                    this.errors.Add(ex);
+                }
+            }
+            else
+            {
+                try
+                {
+                    var tokenResult = await TaskExecutor.CompleteWithin(
+                                    this.logger,
+                                    this.interactiveWindowsAuthTimeout,
+                                    "Get Token Interactive Windows Authentication",
+                                    (cancellationToken) => this.pcaWrapper.GetTokenIntegratedWindowsAuthenticationAsync(this.scopes, cancellationToken),
+                                    this.errors)
+                                    .ConfigureAwait(false);
+                    tokenResult.SetAuthenticationType(AuthType.IntegratedWindowsAuthenticationFlow);
+
+                    return new AuthFlowResult(tokenResult, this.errors);
+                }
+                catch (MsalUiRequiredException ex) when (
+                             ex.Classification == UiRequiredExceptionClassification.BasicAction
+                          && ex.Message.StartsWith("AADSTS50076", StringComparison.OrdinalIgnoreCase))
+                {
+                    this.errors.Add(ex);
+                    this.logger.LogDebug($"IWA failed, 2FA is required.\n{ex.Message}");
+                }
+                catch (MsalUiRequiredException ex) when (
+                             ex.Classification == UiRequiredExceptionClassification.None
+                          && ex.Message.StartsWith("AADSTS500083", StringComparison.OrdinalIgnoreCase))
+                {
+                    this.errors.Add(ex);
+                    this.logger.LogDebug($"IWA failed, unable to verify token signature with identifier 'urn:federation:MSFT'.\n{ex.Message}");

```

Comment:
If this error message is user-exposed, what you do expect a user to do with it? There's not really much to go on if you don't already know the internals (in which case you're probably someone who's working on the tools, or who would rather see the full exception).

Reply:
This is not user-exposed but collected for our telemetry. We queue a bunch of different auth flows depending on the user platform. When we get a null token, we capture the exception for our telemetry and then move on to the next auth flow in the queue.

Reply:
Is catching this error Microsoft specific? I think the first `catch` for ui required exception makes sense, because we can tell the user something that might help. I think this second catch should be an unconditional catch of the MsalUiRequiredException to catch all other forms, and simpel log the message directly, and add it to the error list. We can then track it in Telemetry and see how frequently this comes up. But I agree, it's not actionable to the user, so let's leave it as LogDebug.

Reply:
Updated the second block to catch any type of MSAL UI Exception as it's non-actionable for the users.

Reply:
This error (AADSTS500083) isn't technically Microsoft-specific, but it's specific to a very particular situation (the trust relationship has been broken between the AD and AAD tenants involved) which is pretty uncommon. The changes made seem reasonable.